### ISSUE TITLE
Infinity while using startTimeNS

### ISSUE THREAD

**Issue by QIvan** [OpenHFT/JLBH#4]
Hi! I read through your examples and as far as I understood, we've got to use for sample calculation the parameter `startTimeNS` of the `run` method and I see the same in the javadoc:
``` 
     * @param startTimeNS The time that should be used as the start time for the sample.
```

So, correct me if I'm wrong, in the code, it looks like 
```java
jlbh.sample(System.nanoTime() - startTimeNS);
```
right? 

I made a super simple benchmark, just to be more familiar with JLBH API: 

```java 
import net.openhft.chronicle.core.jlbh.JLBH;
import net.openhft.chronicle.core.jlbh.JLBHOptions;
import net.openhft.chronicle.core.jlbh.JLBHTask;

import java.util.concurrent.locks.LockSupport;

public class SimpleBenchmark implements JLBHTask {

    private JLBH jlbh;

    public static void main(String[] args) {
        //Create the JLBH options you require for the benchmark
        JLBHOptions lth = new JLBHOptions()
                .warmUpIterations(1)
                .iterations(100_000)
                .throughput(100_000)
                .runs(1)
                .jlbhTask(new SimpleBenchmark());
        new JLBH(lth).start();
    }


    @Override
    public void init(JLBH jlbh) {
        this.jlbh = jlbh;
    }

    @Override
    public void run(long startTimeNS) {
//        long start = System.nanoTime();          // (1)
        long start = startTimeNS;                       // (2)
        LockSupport.parkNanos(100_000);

        jlbh.sample(System.nanoTime() - start);
    }
}
```

My expectation was the result is around 100ms but I got this: 
```
SimpleBenchmark.main()
00:48:27.533 [main] DEBUG net.openhft.chronicle.core.Jvm - No system.properties file found
00:48:27.627 [Thread-0] INFO net.openhft.affinity.AffinityLock - No isolated CPUs found, so assuming CPUs 1 to 7 available.
00:48:27.691 [main] INFO net.openhft.affinity.AffinityLock - Assigning cpu 7 to Thread[main,5,main]
00:48:27.704 [main] DEBUG net.openhft.affinity.lockchecker.FileLockBasedLockChecker - Obtained lock on file /tmp/cpu-7.lock (1496)

Warm up complete (1 iterations took 0.167s)
-------------------------------- BENCHMARK RESULTS (RUN 1) --------------------------------------------------------
Run time: 15.378s
Correcting for co-ordinated:true
Target throughput:100000/s = 1 message every 10us
End to End: (100,000)                           50/90 99/99.9 99.99 - worst was -10 / -10  -10 / -10  -10 - -10
OS Jitter (45,571)                              50/90 99/99.9 99.99 - worst was 9.2 / 10  13 / 54  241 - 404
-------------------------------------------------------------------------------------------------------------------
00:48:43.107 [main] INFO net.openhft.affinity.LockInventory - Releasing cpu 7 from Thread[main,5,main]
-------------------------------- SUMMARY (end to end) -----------------------------------------------------------
Percentile   run1      % Variation
50:         Infinity    -Infinity
90:         Infinity    -Infinity
99:         Infinity    -Infinity
99.7:       Infinity    -Infinity
worst:      Infinity    -Infinity
-------------------------------------------------------------------------------------------------------------------
```

but if I comment the line with the label `(1)` and uncomment the line with the label `(2)` the results will be like 
```
SimpleBenchmark.main()
01:00:05.239 [main] DEBUG net.openhft.chronicle.core.Jvm - No system.properties file found
01:00:05.334 [Thread-0] INFO net.openhft.affinity.AffinityLock - No isolated CPUs found, so assuming CPUs 1 to 7 available.
01:00:05.402 [main] INFO net.openhft.affinity.AffinityLock - Assigning cpu 7 to Thread[main,5,main]
01:00:05.415 [main] DEBUG net.openhft.affinity.lockchecker.FileLockBasedLockChecker - Obtained lock on file /tmp/cpu-7.lock (3017)

Warm up complete (1 iterations took 0.171s)
-------------------------------- BENCHMARK RESULTS (RUN 1) --------------------------------------------------------
Run time: 15.44s
Correcting for co-ordinated:true
Target throughput:100000/s = 1 message every 10us
End to End: (100,000)                           50/90 99/99.9 99.99 - worst was 153 / 154  165 / 275  1,590 - 8,720
OS Jitter (39,725)                              50/90 99/99.9 99.99 - worst was 9.3 / 10  36 / 78  306 - 489
-------------------------------------------------------------------------------------------------------------------
01:00:20.881 [main] INFO net.openhft.affinity.LockInventory - Releasing cpu 7 from Thread[main,5,main]
-------------------------------- SUMMARY (end to end) -----------------------------------------------------------
Percentile   run1      % Variation
50:           152.51    -Infinity
90:           154.18    -Infinity
99:           164.67    -Infinity
99.7:         171.71    -Infinity
worst:       8720.38    -Infinity
-------------------------------------------------------------------------------------------------------------------
```

(yes-yes it's my laptop with big OS jittering, but never mind this please)
I'm still not sure is it good that `Variation` is `-Infinity` but the results met my expectations (more or less). 

I use JLBH version 
```java
"net.openhft:jlbh:1.0.1"
```

Is that a bug or it's a mistake in my benchmark? 

Thank you.

**Comment by dpisklov** [comment]
You're blocking the execution for 100 micros in each iteration which means you can't realistically achieve 100_000 iterations/sec, that's the first problem I see here and most likely this is causing other issues - you get infinity because you just don't get all 100_000 iterations completed, and also coordinate omission kicks in. Try reducing throughput, and may be put some more meaningful workload.

**Comment by QIvan** [comment]
hmmm.... interesting, that make sense, thank you. 

Don't you think that jlbh should throw an expection in  such case? Or at least print a warning? 

I saw that the troughput is 10_000 by default, but maybe there is a mode where the task runs the benchmark as it goes (if such mode is even possible)? I mean, that's great that I can specify the throughput, but I can't say that I want to do it for every benchmark. 

thank you.

**Comment by dpisklov** [comment]
@QIvan You need to decide whether you really need JLBH or JMH. JLBH is "taught" how to deal with backing off and fixing coordinate omission - which is the behaviour you see - and that's something very important for the latency benchmark, and moreover it's something it expects to happen, to some extent.
If however you're after microbenchmark (e.g. you just want to know how fast is the method, or a set of methods) - you probably need JMH.

**Comment by QIvan** [comment]
> You need to decide whether you really need JLBH or JMH

for sure, I need jlbh for comparing latency of several transports (like http/grpc/Aeron etc) to understand what еру benifits/penalty of switching form one to another. 
In my benchmark I've got a client and a server and mesure ping pong time between them. My understanding that it's a work for jlbh (correct me if i'm wrong) but my first attempt gave me very strange results that's why i'm here...

**Comment by dpisklov** [comment]
Indeed that's a valid usecase for JLBH.
Can you show us your test and results? We might be able to help.
Also, have a look at [this test](https://github.com/OpenHFT/Chronicle-Queue/blob/master/src/test/java/net/openhft/chronicle/queue/bench/QueueSingleThreadedJLBHBenchmark.java) (and others in that package) - it might provide you with some insights on how we use JLBH.

**Comment by QIvan** [comment]
I couldn't even imagine such an offer! 
I'll clean up my code and share the benchmark. 
Thank you!

**Comment by QIvan** [comment]
hi! sorry that the benchmarks took a bit more time than I expected, but I made it. @dpisklov Could you give me some pointers about what is the right way to use JLBH? 

* like here https://github.com/easy-logic/transport-benchmarks/blob/master/aeron-ping/src/main/java/io/easylogic/benchmarks/AeronPingBenchmarkJlbhSeparateThread.java
* or like here https://github.com/easy-logic/transport-benchmarks/blob/master/aeron-ping/src/main/java/io/easylogic/benchmarks/AeronPingBenchmarkJlbhSeparateThread.java
?
 
I also made a question on StackOverflow (as you recommended me some time ago) where I describe all my concerns: https://stackoverflow.com/questions/63749336/different-results-for-transport-benchmarks

 I tried to make a `JLBH` tag over there, but they say I got to have 1500 reputation at least, so if you're able to moderate the question this might be the first question with that tag )

**Comment by QIvan** [comment]
To the original problem, my understanding with the single-threaded design that infinity result might happen more likely and this is not the right way to use JLBH, am I right? 
thanks!

**Comment by dpisklov** [comment]
@QIvan on the face, I don't see serious issues with the benchmarks from the JLBH perspective. I know nothing about aeron though, so may be there's some issue there.
But the difference in numbers suggests that the coordinated omission is taking place here. My suspect would be the while() loop around publication.offer. Basically in your single-threaded mode, there's no possible message buildup as you send message and them immediately receive it. In multithreaded mode, it's more real-world-like, where it sends messages without checking the other side. That might be causing the difference. I would say the multithreaded test is probably more appropriate, but there's probably some issue in the way you use Aeron.

BTW I tried cloning your repo and running tests - none of them runs successfully for me (they complain about media driver, and if I run media driver separately nothing happens at all - even warmup won't complete). I don't run those in Docker however, just run locally in IDE. In general, I wouldn't advise running perf tests in docker...

As for the infinity, if the benchmark is correct you should never get infinity.

**Comment by peter-lawrey** [comment]
The test tries to start every 10 micro-seconds but its taking far longer resulting in the test getting further and further behind. If you reduce the throughput to say 5,000/s so it has enough time to perform a task that takes over 100 microseconds I get the following.

```java
        JLBHOptions lth = new JLBHOptions()
                .warmUpIterations(1)
                .iterations(100_000)
                .throughput(5_000)
//                .accountForCoordinatedOmmission(false)
                .runs(2)
                .jlbhTask(new OneResultBenchmark());
        new JLBH(lth).start();
```

prints

```
[main] INFO net.openhft.affinity.LockInventory - Releasing cpu 11 from Thread[main,5,main]
Percentile   run1         run2      % Variation
50:           152.06       152.06         0.00
90:           154.11       155.14         0.00
99:          1323.01      9863.17         0.00
99.7:        4055.04     22085.63         0.00
worst:       9601.02     32309.25         0.00
```

Or you can ignore co-ordinated omission which means that each test expects to wait for the previous task to finish before starting the next one. This can produce unrealistically good result.

```java
        JLBHOptions lth = new JLBHOptions()
                .warmUpIterations(1)
                .iterations(100_000)
                .throughput(100_000)
                .accountForCoordinatedOmission(false)
                .runs(2)
                .jlbhTask(new OneResultBenchmark());
        new JLBH(lth).start();
```

```
Run time: 16.275s, distribution: NORMAL
...
Percentile   run1         run2      % Variation
50:           152.06       152.06         0.00
90:           152.06       152.06         0.00
99:           156.16       155.14         0.00
99.7:         162.30       159.23         0.00
worst:      13926.40      3874.82         0.00
```

NOTE: This should have completed in 10 seconds but actually took 16.

----

For comparison you may find this interesting.

```java
public class SimpleBenchmark implements JLBHTask {

    private JLBH jlbh;

    public static void main(String[] args) {
        //Create the JLBH options you require for the benchmark
        JLBHOptions lth = new JLBHOptions()
                .warmUpIterations(20_000)
                .iterations(100_000)
                .throughput(10_000)
                .accountForCoordinatedOmission(true)
                .runs(2)
                .jlbhTask(new SimpleBenchmark());
        new JLBH(lth).start();
    }


    @Override
    public void init(JLBH jlbh) {
        this.jlbh = jlbh;
    }

    @Override
    public void run(long startTimeNS) {
//        long start = System.nanoTime();          // (1)
        long start = startTimeNS;                       // (2)
        LockSupport.parkNanos(1);

        final long delta = System.nanoTime() - start;
        jlbh.sample(delta);
    }
}
```

prints on my machine

```
Warm up complete (20000 iterations took 1.132s)
-------------------------------- BENCHMARK RESULTS (RUN 1) --------------------------------------------------------
Run time: 10.0s, distribution: NORMAL
Correcting for co-ordinated:true
Target throughput:10000/s = 1 message every 100us
End to End: (100,000)                           50/90 99/99.9 99.99 - worst was 52.1 / 52.9  66.8 / 4,310  6,900 - 7,290
OS Jitter (7,897)                               50/90 99/99.9 99.99 - worst was 16.1 / 23.4  31.6 / 45.2  137 - 137
-------------------------------------------------------------------------------------------------------------------
-------------------------------- BENCHMARK RESULTS (RUN 2) --------------------------------------------------------
Run time: 10.0s, distribution: NORMAL
Correcting for co-ordinated:true
Target throughput:10000/s = 1 message every 100us
End to End: (100,000)                           50/90 99/99.9 99.99 - worst was 52.1 / 53.1  61.1 / 136  752 - 1135
OS Jitter (5,868)                               50/90 99/99.9 99.99 - worst was 16.0 / 23.9  32.1 / 67.8  99.6 - 99.6
-------------------------------------------------------------------------------------------------------------------
-------------------------------- SUMMARY (end to end) -----------------------------------------------------------
Percentile   run1         run2      % Variation
50:            52.10        52.10         0.00
90:            52.86        53.12         0.00
99:            66.82        61.06         0.00
99.7:         579.58        77.06         0.00
worst:       7290.88      1134.59         0.00
-------------------------------------------------------------------------------------------------------------------
```

**Comment by peter-lawrey** [comment]
@QIvan To address the original question, there was an assumption that latencies would be less than 4.2 seconds or they overflowed. The `Histogram` has a default limit of 1.2 hours and this is a safer limit. I have changed the code so even for extreme delays you still see a result.

```
Warm up complete (20000 iterations took 1.136s)
-------------------------------- BENCHMARK RESULTS (RUN 1) --------------------------------------------------------
Run time: 52.595s, distribution: NORMAL
Correcting for co-ordinated:true
Target throughput:100000/s = 1 message every 10us
End to End: (1,000,000)                         50/90 99/99.9 99.99/99.999 - worst was 21,139,290 / 38,252,050  42,278,580 / 42,547,020  42,547,020 / 42,547,020 - 42,547,020
OS Jitter (34,930)                              50/90 99/99.9 99.99 - worst was 16.2 / 23.6  33.2 / 43.6  97.5 - 1,290
-------------------------------------------------------------------------------------------------------------------
-------------------------------- BENCHMARK RESULTS (RUN 2) --------------------------------------------------------
Run time: 52.513s, distribution: NORMAL
Correcting for co-ordinated:true
Target throughput:100000/s = 1 message every 10us
End to End: (1,000,000)                         50/90 99/99.9 99.99/99.999 - worst was 21,407,730 / 38,252,050  42,010,150 / 42,547,020  42,547,020 / 42,547,020 - 42,547,020
OS Jitter (47,919)                              50/90 99/99.9 99.99 - worst was 1061 / 1094  1110 / 1143  1,390 - 4,770
-------------------------------------------------------------------------------------------------------------------
-------------------------------- SUMMARY (end to end) -----------------------------------------------------------
Percentile   run1         run2      % Variation
50:      21139292.16  21407727.62         0.00
90:      38252052.48  38252052.48         0.00
99:      42278584.32  42010148.86         0.00
99.7:    42547019.78  42278584.32         0.00
99.9:    42547019.78  42547019.78         0.00
worst:   42547019.78  42547019.78         0.00
-------------------------------------------------------------------------------------------------------------------
```

**Comment by hft-team-city** [comment]
Released in [JLBH-1.21ea2](https://github.com/OpenHFT/JLBH/releases/jlbh-1.21ea2), [BOM-2.21ea125](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.21ea125)
